### ISSUE TITLE
Project abandoned?

### ISSUE THREAD

**Issue by nnarhinen** [TTLabs/EvaporateJS#109]
@tomsaffell are you planning on reviewing / merging the pull requests by @bikeath1337 ? Seems they address some pretty useful features..

@bikeath1337 do you have all your fixes (those regarding resuming, especially) in one branch or are they all in feature branches?

**Comment by bikeath1337** [comment]
@nnarhinen @tomsaffell has not replied since very early 2016, in spite of personal offers to assist. I'm considering this project abandoned.

**Comment by tomsaffell** [comment]
@nnarhinen @bikeath1337 Please accept my apologies for being checked out! I've been exceptionally busy, and I feel guilty for not checking in for so long. Looks like some great progress has been made!

Do either of you guys have suggestions about how to solve this? I could accept all the PRs and hope it all works out. I could give someone else admin control of the repo (I presume github allows that, but I've never looked). I guess we could mark the current master head as a 'release', then accept all the PRs as a 'development' version. Best to do that on a branch, or master? What other options are there? 

Sorry, I'm not experienced running an OS project, so I'm not sure what the norm is..

tom

**Comment by rickihastings** [comment]
I think the best thing to do is tag current master as a release, start merging pull requests into development, test. And then bump the version as needed etc.

I'd like to see this project more maintained as I'm thinking about using it on a large scale project but I'm not sure how stable it currently is?

**Comment by tomsaffell** [comment]
@rickihastings that sounds like a good suggestion - thanks. 

in terms of stability, I think the current master head is very stable. we've been using it in production for 2 years, and we've seen 100GB files uploaded with it. if you have any specific stability questions then please post them as a new issue and we'll discuss them there.

**Comment by bikeath1337** [comment]
With all good intentions, the version at master is  not stable at all under stress. Primarily, it has the following problems and shortcomings:

1. a huge memory leak cause the the retry logic. This can be seen by disconnecting the internet and then letting EvaporateJS retry for an hour or so. This is because references to the aborted xhrs are maintained in a list, thus never garbage collected. There is a PR i wrote for this.
2. when the user aborts the upload, EvaporateJS does not notify AWS to clear the multipart upload resources. This is a problem because those resources are charged back to the bucket owner and are not visible except through API calls to fetch incomplete multipart uploads. I submitted a PR for this.
3. the Initiate and Complete steps do not retry on failure. Amazon AWS explicitly states that all S3 API calls can fail (with a 500 response). And that the application is expected to retry until success.

We saw 3. happen in our product environment, which resulted in complete failure of the upload after a long wait (1.5GB) only because the Complete method failed. I also submitted a PR for this.

At this point, we've gone to production with a complete and utterly robust solution and great success. Some of the pull requests tread on each other a bit, so merging requires careful attention. I purposefully attempted to follow existing patterns, even though they could have been improved. This was to make it easier to validate the changes and focus just on what was changed.

**Comment by tomsaffell** [comment]
Thanks @bikeath1337 for those clarifications. You've certainly spent more time with your head in this code recently than i have, so you are the better judge of stability. All three points are valid, though I wouldnt say (2) is a matter of stability.

FYI to anyone dealing with cleaning up a repo with containing failed uploads (as caused by (2)): the AWS CLI can help clean them up, as can an AWS library such as boto (python), which is what we use. Still, it's makes sense for EvaporateJS to do it, so thanks for that PR.

I've never seen (3), but that's probably due to being blind to it, so again, thanks. 

I'll get your PRs merged this coming week. Thanks for following existing patterns - I agree they could be better - honestly, I wrote this code in a couple of days three years ago, and havent really spent much time on it since. @bikeath1337 would you like to have admin access to the repo? Seems like you're closer to it than me right now.  Also, do you suggest that I accept all your PRs together, or is there a subset that would result in the most stable code? Thanks again for your contributions

@rickihastings hope this helps answer your question regarding stability.

**Comment by rickihastings** [comment]
This is good news to hear that the library is used in production. I'll be giving it a shot and seeing how things go. I would like to see those 3 merge requests merged in though as they're great improvements.

**Comment by bikeath1337** [comment]
No problem. The biggest addition is the one related to checksums. As I wrote to you back in December, not even the AWS Javascript API for Browsers supports the checksums, which is essential for fidelity of copying. Amazingly, the library does support the checksum but only for string bodies, not for huge blobs like video, which is exactly the use case that needs the feature!

I would be more than happy to create a merge candidate if you want to give me admin status. (I’m an admin to the ruby gem streamio-ffmpeg already).

Bobby

**Comment by tomsaffell** [comment]
@bikeath1337 i made you an admin. do you need me to do anything else right now?

**Comment by bikeath1337** [comment]
Thank you @tomsaffell . That should be good for now. I will not take any actions on branches without the usual feedback. I have a branch ready to go but as I did it last week, I would like to check it out again. I think you will be impressed with the end result.

**Comment by swangui** [comment]
@bikeath1337 Thanks for pointing out the cost related issues. Could you please add a section in the README.md so that people like me will notice this after paying for s3 a while already. Would be great if you can list the API for people to lookup and clean incomplete multi-parts. Thanks!

**Comment by bikeath1337** [comment]
In https://github.com/TTLabs/EvaporateJS/pull/111 I added text to the readme regarding how to clean up uploads. I also included a link to a Rails rake task that works for this purpose.

The thing to keep in mind is that ANY access to S3 could fail. Amazon documents this. So...unless the script automatically retries, if you get a 500 response from S3, you just have to try again until it succeeds. 

My rake task does not retry, so I just run it again until it succeeds.
